lib/mirage/mirage.mli

```diff
@@ -15,19 +15,103 @@
+    container configured with a tight security profile. In general, these
+    platform do not have a full POSIX environment; MirageOS does not try to
+    emulate POSIX and focuses on providing a small, well-defined, typed
+    interface with the system components. The nearest equivalent to the MirageOS
+    approach is the WASI (wasi.dev) set of interfaces for WebAssembly.
+
+    {2 Is everything really written in OCaml?}
+
+    While most of the code is written in OCaml, a typed, high-level language
+    with many good safety properties, there are pieces of MirageOS which are
+    still written in unsafe C. These bits can be separated in three categories:
+
+    - The OCaml runtime is written in C. It needs to be ported to the platform
+      that MirageOS is trying to target, which do not support POSIX. Hence, the
+      first component to port to a new platform is the OCaml runtime.
+
+    - The low-level device drivers (network, console, clock, etc) also need some
+      C bits.
+
+    - The base usual C bindings; some libraries are widely used and
+      (unfortunately) very hard (but not impossible) to replace them completely
+      without taking a big performance hit or having to trust code without much
+      real-world usages. This is the case for low-level bit handling for crypto
+      code (even if we try to make sure allocation is alway handled by the OCaml
+      runtime) as well as arbitrary precision numeric computation (e.g. gmp).
+      Ideally we could image rewriting all of these libraries in OCaml if we had
+      an infinite amount of time in our hands.
+
+    {2 MirageOS as a cross-compilator}
+
+    The MirageOS compiler is basically a cross-compiler, where the host and
+    target toolchain are identical, but with different flags for the C bindings:
+    for instance, it is necessary to pass [-freestanding] to {e all} C bindings
+    to not use POSIX headers. The MirageOS compiler also uses a custom linker:
+    eg. not only it needs a custom OCaml's runtime [libasmrun.a], but it also
+    needs to run a different linker to generate specialised executable images.
+
+    Historically, the OCaml ecosystem always had partial support for
+    cross-compilation: for instance, the
+    {{:https://github.com/ocaml-cross/opam-cross-windows} ocaml-cross} way of
+    doing it is to duplicate {e all} existing opam pacakges by adding a
+    [-windows] suffix to their names and dependencies; this allows normal
+    packages and windows packages can be co-installed in the same opam switch.
+
+    MirageOS 3 solves this by duplicating only the packages defining C bindings
+    and (ab)using
+    {{:https://github.com/ocaml/opam-repository/blob/master/packages/zarith-xen/zarith-xen.1.7/files/mirage-install.sh#L20}
+    ocamlfind predicates} to resolve link time depencencies for C archives and
+    on relying on the now deprecated [--output-obj] option of the OCaml
```

Comment:
Currently, `--output-obj` is not really deprecated where a discussion still exists on this specific point here: https://github.com/ocaml/ocaml/pull/1351. Then, the `dune` team decided to use `--output-complete-obj` https://github.com/ocaml/dune/pull/23 in the same time.